### PR TITLE
Currency limits and validation

### PR DESCRIPTION
**Pull request by porteron**
- Included a limit to the number of currencies and tokens allowed to be
sent in single request

- Added in a known currencies list for coingecko, we will filter based
upon this list. Should save some requests. List pulled from here:
https://api.coingecko.com/api/v3/simple/supported_vs_currencies

- Ran linter

### CODE REVIEW

File: versions/v2/coingecko.js [hunk 1]
```diff
@@ -8,6 +8,9 @@ const cache = Cache.create(5 * 60, {
   url: env.REDIS_URL
 })
```
Comment (porteron) [inline versions/v2/coingecko.js:10]: limits here

File: versions/v2/coingecko.js [hunk 2]
```diff
@@ -23,6 +26,70 @@ module.exports = {
   fetchCoinsList
 }
```
Comment (porteron) [inline versions/v2/coingecko.js:28]: List of supported currencies

File: versions/v2/coingecko.js [hunk 3]
```diff
@@ -145,50 +212,64 @@ async function generateMappings (coinlist) {
     yld: 'yield',
     zap: 'zap'
   }
-  return payload.reduce((memo, { id, symbol, platforms }) => {
-    if (special[symbol] && id !== special[symbol]) {
+  return payload.reduce(
```
Comment (porteron) [inline versions/v2/coingecko.js:215]: Lint fix

File: versions/v2/coingecko.js [hunk 4]
```diff
@@ -145,50 +212,64 @@ async function generateMappings (coinlist) {
...
+  )
+}
 
-    if (id !== 'link') {
-      memo.idToSymbol[id] = symbol
-    }
-    memo.symbolToId[symbol] = id
-    if (platforms && platforms.ethereum) {
-      memo.hashToId[platforms.ethereum.toLowerCase()] = id
+async function fetchCoinsList (
```
Comment (porteron) [inline versions/v2/coingecko.js:238]: Lint fix

File: versions/v2/coingecko.js [hunk 5]
```diff
@@ -145,50 +212,64 @@ async function generateMappings (coinlist) {
...
 }
 
-async function fetchCoinsList (nil, {
-  platform // boolean
-}) {
-  const result = await passthrough({}, {
-    path: '/api/v3/coins/list',
-    query: {
-      include_platform: platform
+function filterSupportedCurrencies (currencies = []) {
```
Comment (porteron) [inline versions/v2/coingecko.js:256]: function to filter unsupported currencies

File: versions/v2/coingecko.js [hunk 6]
```diff
@@ -145,50 +212,64 @@ async function generateMappings (coinlist) {
...
 
-async function rates ({
-  a,
-  b,
-  from,
-  until
-}, {
-  refresh
-}) {
+async function rates ({ a, b, from, until }, { refresh }) {
```
Comment (porteron) [inline versions/v2/coingecko.js:266]: Lint fix

File: versions/v2/coingecko.js [hunk 7]
```diff
@@ -145,50 +212,64 @@ async function generateMappings (coinlist) {
...
   const [a1, b1] = await mapIdentifiers(a, b)
 
   const lowerFrom = from.toLowerCase()
-  let f = knownTimeWindows[lowerFrom] ? await knownTimeWindows[lowerFrom]() : from
+  let f = knownTimeWindows[lowerFrom]
```
Comment (porteron) [inline versions/v2/coingecko.js:270]: lint fix

File: versions/v2/coingecko.js [hunk 8]
```diff
@@ -197,172 +278,195 @@ async function rates ({
   u = toSeconds(u)
 
   const a1Id = a1.map(({ id }) => id).join(',')
+  let vsCurrency = b1.map(({ symbol: id }) => id)
```
Comment (porteron) [inline versions/v2/coingecko.js:281]: Filter currencies

File: versions/v2/coingecko.js [hunk 9]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
-      path: `/api/v3/coins/${a1Id}/market_chart/range`,
-      query: {
-        vs_currency: b1.map(({ symbol: id }) => id).join(','),
-        from: truncate5Min(f),
-        to: truncate5Min(u)
+    if (!vsCurrency.length) {
+      throw new Error({ message: 'Currencies not supported' })
+    }
+
+    const result = await passthrough(
```
Comment (porteron) [inline versions/v2/coingecko.js:289]: lint fix

File: versions/v2/coingecko.js [hunk 10]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
+
   try {
-    const result = await passthrough({}, {
-      refresh,
-      path: `/api/v3/coins/${a1Id}/market_chart/range`,
-      query: {
-        vs_currency: b1.map(({ symbol: id }) => id).join(','),
-        from: truncate5Min(f),
-        to: truncate5Min(u)
+    if (!vsCurrency.length) {
```
Comment (porteron) [inline versions/v2/coingecko.js:285]: Check to see if there is at least one supported currency

File: versions/v2/coingecko.js [hunk 11]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
 const knownTimeWindows = {
-  live: () => (new Date()) - hour1,
-  '1h': () => (new Date()) - hour1,
-  '1d': () => (new Date()) - day1,
-  '1w': () => (new Date()) - week1,
-  '1m': () => (new Date()) - month1,
-  '3m': () => (new Date()) - (month1 * 3),
-  '1y': () => (new Date()) - year1,
-  all: () => +(new Date('2014-01-01'))
+  live: () => new Date() - hour1,
```
Comment (porteron) [inline versions/v2/coingecko.js:308]: lint fix

File: versions/v2/coingecko.js [hunk 12]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
 
-async function spotPrice ({
-  a,
-  b,
-  from,
-  until
-}, {
-  refresh
-}) {
+async function spotPrice ({ a, b, from, until }, { refresh }) {
```
Comment (porteron) [inline versions/v2/coingecko.js:318]: lint fix

File: versions/v2/coingecko.js [hunk 13]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
   let ratesResult = null
   const [a1, b1] = await mapIdentifiers(a, b)
 
   const a1Id = a1.map(({ id }) => id).join(',')
   const b1Id = b1.map(({ symbol: id }) => id).join(',')
   if (from || until) {
     const lowerFrom = from.toLowerCase()
-    const f = knownTimeWindows[lowerFrom] ? await knownTimeWindows[lowerFrom]() : from
+    const f = knownTimeWindows[lowerFrom]
```
Comment (porteron) [inline versions/v2/coingecko.js:326]: lint fix

File: versions/v2/coingecko.js [hunk 14]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
-          from: f_ + ''
-        }
-        if (u) {
-          arg1.until = u + ''
-        }
-        const arg2 = {
-          refresh
-        }
-        const { payload = { prices: [] } } = await rates(arg1, arg2)
+      const subRate = (ratesResult[a] = ratesResult[a] || {})
```
Comment (porteron) [inline versions/v2/coingecko.js:333]: lint fix

File: versions/v2/coingecko.js [hunk 15]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
     }
   }
 
-  const result = await passthrough({}, {
-    refresh,
-    path: '/api/v3/simple/price',
-    query: {
-      ids: a1Id,
-      vs_currencies: b1Id
+  const result = await passthrough(
```
Comment (porteron) [inline versions/v2/coingecko.js:369]: lint fix

File: versions/v2/coingecko.js [hunk 16]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
+      }
     }
-  })
+  )
 
-  result.payload = _.reduce(result.payload, (memo, value, a) => {
-    const { symbol = '', id = '', address = null } = _.find(a1, ({ id, symbol }) => id === a || symbol === a)
-    const subTarget = ratesResult && (ratesResult[symbol] || ratesResult[id] || ratesResult[address])
-    const subList = subTarget ? Object.keys(subTarget) : []
+  result.payload = _.reduce(
```
Comment (porteron) [inline versions/v2/coingecko.js:381]: lint fix

File: versions/v2/coingecko.js [hunk 17]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
-        }
-        const current = new BigNumber(value[k])
-        const subRate = subTarget[k]
-        if (!subRate) return
-        const previous = new BigNumber(subRate[1])
-        const deltaKey = `${k}_timeframe_change`
-        value[deltaKey] = current.minus(previous).dividedBy(previous).times(100).toNumber()
-      })
-    })();
+      // For each currency set timeframe property
```
Comment (porteron) [inline versions/v2/coingecko.js:397]: lint fix

File: versions/v2/coingecko.js [hunk 18]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
-    (function setTokenValue () {
-      a1.forEach(({ id, symbol, converted }) => {
-        if (symbol !== a && id !== a) {
-          return
-        }
-        if (converted.symbolToId) {
-          memo[symbol] = value
-          if (id !== symbol) {
-            delete memo[id]
+      // For each token compare w/ value returned from rates and update memo
```
Comment (porteron) [inline versions/v2/coingecko.js:420]: lint fix

File: versions/v2/coingecko.js [hunk 19]
```diff
@@ -197,172 +278,195 @@ async function rates ({
...
+      (function setTokenValue () {
+        a1.forEach(({ id, symbol, converted }) => {
+          if (symbol !== a && id !== a) {
+            return
           }
-        }
-      })
-    })()
+          if (converted.symbolToId) {
```
Comment (porteron) [inline versions/v2/coingecko.js:426]: lint fix

File: versions/v2/coingecko.js [hunk 20]
```diff
@@ -379,11 +483,7 @@ async function mapIdentifiers (...currencies) {
 }
 
 // second argument is a parsed query string
-function passthrough (notvar, {
-  refresh,
-  path: basePath,
-  query
-}) {
+function passthrough (notvar, { refresh, path: basePath, query }) {
```
Comment (porteron) [inline versions/v2/coingecko.js:486]: lint fix

File: versions/v2/coingecko.js [hunk 21]
```diff
@@ -393,30 +493,36 @@ function passthrough (notvar, {
...
     let ret = +d
-    if (ret > ((new Date() / 1000) + 1000)) {
+    if (ret > new Date() / 1000 + 1000) {
       ret = ret / 1000
     }
     return parseInt(ret)
   }
   const date = new Date(d)
-  if (date.getYear() < 100) { // already in seconds format
+  if (date.getYear() < 100) {
```
Comment (porteron) [inline versions/v2/coingecko.js:524]: lint fix

File: versions/cache.js [hunk 22]
```diff
@@ -51,7 +51,7 @@ const mock = {
 }
 
 // increase to 5 min cache
-function create (expiry = 60*5, options) {
+function create (expiry = 60 * 5, options) {
```
Comment (porteron) [inline versions/cache.js:54]: lint fix

**Review by husobee — APPROVED** [review]
lgtm

**Pull request merged**